File: frontend/src/interpreter/parser.ts
import {
  AssignNode,
  AST,
  BinaryOperatorNode,
  CompoundNode,
  ComputableNode,
  isComputableNode,
  SingleValueNode,
  UnaryOperatorNode,
  VariableNode,
  ProcedureNode,
} from "./ast";
import { Matrix, AComputable } from "./computable";
import { MatrixError, ParsingError, SymbolError, SyntaxError } from "./errors";
import { Lexer } from "./lexer";
import {
  TokenType,
  Token,
  isNumericToken,
  plus_token,
  eof_token,
  isSymbolToken,
} from "./token";

export class Parser {
  private lexer: Lexer;
  private current_token: Token;

  constructor(lexer: Lexer) {
    this.lexer = lexer;
    this.current_token = this.lexer.next_token();
  }

  /**
   * identify expression
   * expr   : term ((PLUS | MINUS) term)* | term ((PLUS | MINUS)term)*
   * term   : factor ((MUL | DIV | POW) factor)*
   * factor : (PLUS | MINUS) factor | NUMBER | lparen expr rparen | matrix | variable | procedure
   * matrix :  lbracket (row)* rbracket
   */
  private expr(ignoreWhiteSpace: boolean = true): AST {
    // first node
    let node: AST = this.term();

    // do this until a new ((PLUS | MINUS) term) cannot be found
    while (
      this.current_token.type == TokenType.plus ||
      this.current_token.type == TokenType.minus
    ) {
      // check white spaces,
      // if next character is empty or previous character is empty, binary operate
      if (
        ignoreWhiteSpace ||
        this.lexer.peek(1) == " " ||
        this.lexer.peek(-1) != " "
      ) {
        // capture operator (PLUS | MINUS)
        let token = this.current_token;
        if (token.type == TokenType.plus) {
          this.eat(TokenType.plus);
        } else if (token.type == TokenType.minus) {
          this.eat(TokenType.minus);
        }
        // build abstract syntax tree
        node = new BinaryOperatorNode(node, token, this.term());
      } else {
        return node;
      }
    }

    return node;
  }

  /**
   * identify term
   * term   : factor ((MUL | DIV) factor)*
   * factor : (PLUS | MINUS) factor | NUMBER | lparen expr rparen | matrix
   * matrix :  lbracket (row)* rbracket
   * @return node
   */
  private term(): AST {
    // left factor
    let node: AST = this.powers();
    // do this until a new ((POW | MUL | DIV | RDIV) factor) cannot be found
    while (
      this.current_token.type == TokenType.mul ||
      this.current_token.type == TokenType.div ||
      this.current_token.type == TokenType.rdiv
    ) {
      // capture next operator (MUL | DIV)
      let token = this.current_token;
      if (token.type == TokenType.mul) {
        this.eat(TokenType.mul);
      } else if (token.type == TokenType.div) {
        this.eat(TokenType.div);
      } else if (token.type == TokenType.rdiv) {
        this.eat(TokenType.rdiv);
      }
      // build abstract syntax tree
      node = new BinaryOperatorNode(node, token, this.powers());
    }

    return node;
  }

  private powers(): AST {
    // left factor
    let node: AST = this.factor();
    while (this.current_token.type == TokenType.pow) {
      let token = this.current_token;
      if (token.type == TokenType.pow) {
        this.eat(TokenType.pow);
      }
      node = new BinaryOperatorNode(node, token, this.factor());
    }
    return node;
  }

  /**
   * identify factor
   * factor : (PLUS | MINUS) factor | NUMBER | lparen expr rparen | matrix
   * matrix :  lbracket (row)* rbracket
   * @return abstract syntax node
   */
  private factor(): AST {
    let token = this.current_token;
    if (token.type == TokenType.plus) {
      // positive sign
      this.eat(TokenType.plus);
      return new UnaryOperatorNode(token, this.factor());
    } else if (token.type == TokenType.minus) {
      // negation sign
      this.eat(TokenType.minus);
      return new UnaryOperatorNode(token, this.factor());
    } else if (isNumericToken(token)) {
      // token is a number
      this.eat(TokenType.num);
      return new SingleValueNode(token.value);
    } else if (token.type == TokenType.lparen) {
      // lparen expr rparen
      this.eat(TokenType.lparen);
      let node: AST = this.expr();
      this.eat(TokenType.rparen);
      return node;
    } else if (token.type == TokenType.lbracket) {
      // matrix
      let matrix: Matrix = this.matrix();
      return new SingleValueNode(matrix);
    } else if (token.type == TokenType.larrow) {
      let vector: Matrix = this.vector();
      return new SingleValueNode(vector);
    } else if (token.type == TokenType.id) {
      let next = this.lexer.peek();
      if (next && next == "(") {
        return this.procedure();
      } else {
        return this.variable();
      }
    }

    throw new SyntaxError("unexpected symbol");
  }

  private vector(): Matrix {
    // check left arrow
    this.eat(TokenType.larrow);
    // get one row
    let row = this.matrix_row(TokenType.rarrow);
    // check right arrow
    this.eat(TokenType.rarrow);
    return new Matrix([row]);
  }

  /**
   * matrix :  lbracket (row ;)* (row ]) rbracket
   */
  private matrix(): Matrix {
    let arr: ComputableNode[][] = new Array();
    // check left bracket
    this.eat(TokenType.lbracket, "parsing matrix: ");
    // loop through rows
    while (this.current_token.type != TokenType.rbracket) {
      let row = this.matrix_row(TokenType.rbracket);
      arr.push(row);
      if (this.current_token.type == TokenType.semicolon) {
        this.eat(TokenType.semicolon, "parsing matrix row: ");
      }
    }
    // check right bracket
    this.eat(TokenType.rbracket, "parsing matrix: ");
    return new Matrix(arr);
  }

  /**
   * row : (factor,)*
   */
  private matrix_row(endToken: TokenType): ComputableNode[] {
    let arr: ComputableNode[] = new Array();

    while (true) {
      let val = this.expr(false);
      // make sure the element is computable
      if (isComputableNode(val)) {
        // everything else is good, push element into row
        arr.push(val);
        // if row hasn't reached end, consume 'comma' separator
        if (
          this.current_token.type != TokenType.semicolon &&
          this.current_token.type != endToken
        ) {
          if (this.current_token.type == TokenType.comma) {
            this.eat(TokenType.comma, "parsing matrix row: ");
          }
        } else {
          break;
        }
      } else {
        throw new MatrixError(
          "matrix parsing error: expected a numeric element"
        );
      }
    }

    return arr;
  }

  /**
   * consume token and advance to next token
   * @param type token type to verify
   */
  private eat(type: TokenType, message?: string) {
    if (this.current_token.type == type) {
      // if expected token type and current token type matches, proceed to next token
      this.current_token = this.lexer.next_token();
    } else {
      // if token type does not match, a syntax error has happened
      throw new SyntaxError(
        message ||
          "" +
            "expected " +
            type.toString() +
            ", but got " +
            this.current_token.type.toString()
      );
    }
  }

  /**
   * variable : id
   */
  public variable(): VariableNode {
    if (isSymbolToken(this.current_token)) {
      let node = new VariableNode(this.current_token);
      this.eat(TokenType.id);
      return node;
    }
    throw new SymbolError("couldn't parse variable name");
  }

  /**
   * assignment : (variable = expr) (,variable = expr)*
   */
  public assignment(): AssignNode[] {
    let left = this.variable();
    let token = this.current_token;
    this.eat(TokenType.assign, "parsing assignment: ");
    let right = this.expr();
    let assignments = [new AssignNode(left, token, right)];
    //  check if multiline assignment
    if (this.current_token.type == TokenType.comma) {
      this.eat(TokenType.comma);
      let next = this.assignment();
      assignments = assignments.concat(next);
    }
    return assignments;
  }

  /**
   * procedure : id lparen (expr,)* rparen
   */
  public procedure(): ProcedureNode {
    let token = this.current_token;
    if (!isSymbolToken(token)) {
      throw new SymbolError("expected an identifier symbol but didn't get it!");
    }
    this.eat(TokenType.id);
    this.eat(TokenType.lparen);
    let args = [];
    while (true) {
      args.push(this.expr());
      if (this.current_token.type == TokenType.comma) {
        this.eat(TokenType.comma);
      } else {
        break;
      }
    }
    this.eat(TokenType.rparen);
    return new ProcedureNode(token, args);
  }

  /**
   * assignment | procedure | expr
   */
  public variable_statement(): AST[] {
    let token = this.current_token;
    if (token.type == TokenType.id) {
      if (this.lexer.peekToken() == "=") {
        return this.assignment();
      } else if (this.lexer.peekToken() == "(") {
        return [this.procedure()];
      } else {
        return [this.expr()];
      }
    }

    throw new ParsingError("couldn't find a variable!");
  }

  /**
   * statement : compound_statement | assignment_statment | expression_statement | empty
   */
  public statement(): AST[] {
    if (this.current_token.type == TokenType.id) {
      return this.variable_statement();
    } else {
      return [this.expr(true)];
    }
  }

  /**
   * statement_list : statement | statement endl statement_list
   */
  public statement_list(): AST[] {
    let results = this.statement();
    while (this.current_token.type == TokenType.endl) {
      while (this.current_token.type == TokenType.endl) {
        this.eat(TokenType.endl);
      }

      if (this.current_token.type != TokenType.eof) {
        results = results.concat(this.statement());
      }
    }

    if (this.current_token.type == TokenType.id) {
      throw new SyntaxError("unexpected identifier");
    }

    return results;
  }

  /**
   * compound: statement_list
   */
  public compound(): AST {
    let nodes = this.statement_list();

    let node = new CompoundNode();
    for (let n of nodes) {
      node._children.push(n);
    }

    return node;
  }

  public program(): AST {
    let node = this.compound();
    this.eat(TokenType.eof);
    return node;
  }

  /**
   * parse tokens into an abstract syntax tree for traversal
   */
  public parse(): AST {
    let node = this.program();
    if (this.current_token.type != TokenType.eof) {
      throw new SyntaxError("parsing didn't go as expected!");
    }
    return node;
  }
}

File: interpreter/src/parser.rs
use crate::lexer::{TokenType, TokenValue, Lexer, Token};
use crate::errors::Error;
use crate::ast::{Expr, Block, Empty, Assignment, Variable, Binary, Ternary, Unary, Value, FnCall, FnDecl, Sequence, MatrixExpr};
use crate::lexer::TokenType::{Eof, EndL, Id, Reserved, Assign, Comma, Plus, Colon, Minus, NotUnary, Num, LParen, RParen, Def, LBrace, RBrace, LBracket, RBracket, Semicolon, ReservedValue};
use crate::data::Data;

const OP_LEVELS: i8 = 8;

pub struct Parser {
    lexer: Lexer,
    current: Token,
}

impl Parser {
    pub fn new(text: &str) -> Parser {
        let mut lexer = Lexer::new(text);
        let current = lexer.next_token().unwrap();
        return Parser { lexer, current };
    }

    pub fn parse(&mut self) -> Result<Expr, Error> {
        let node = self.program();
        if self.current.token != TokenType::Eof {
            return Err(Error::WrongType("expected eof".to_string()));
        }
        return node;
    }

    fn program(&mut self) -> Result<Expr, Error> {
        let node = self.block();
        self.eat(Eof);
        return node;
    }

    fn block(&mut self) -> Result<Expr, Error> {
        let mut output = Vec::new();
        loop {
            self.eat_new_lines();
            let statement = self.statement().unwrap();
            output.push(statement);

            if self.current.token != EndL {
                // statement is broken
                break;
            }
        }
        return Ok(Expr::Block(Box::new(Block { children: output })));
    }

    fn statement(&mut self) -> Result<Expr, Error> {
        return if self.current.token == Id {
            self.identifier()
        } else if self.current.token == Reserved {
            self.reserved()
        } else if self.current.token != Eof {
            self.expr(true)
        } else {
            Ok(Expr::Empty(Empty {}))
        };
    }

    fn identifier(&mut self) -> Result<Expr, Error> {
        return if self.current.token == Id {
            if self.lexer.peek_token().unwrap().token == Assign {
                self.assignments()
            } else {
                self.expr(true)
            }
        } else {
            Err(Error::IdentifierNotFound(String::from("incorrect identifier")))
        };
    }

    fn reserved(&mut self) -> Result<Expr, Error> {
        let val = &self.current.value;

        if let TokenValue::String(val) = val {
            return if val == "true" {
                self.expr(true)
            } else if val == "false" {
                self.expr(true)
            } else {
                Err(Error::WrongType(String::from("Expected a boolean")))
            };
        }

        return Err(Error::WrongType(String::from("Expected a boolean")));
    }

    fn rec_binary(&mut self, level: i8, ignore: bool) -> Result<Expr, Error> {
        if level < 0 {
            return Ok(self.factor(ignore).unwrap());
        }

        let mut node = self.rec_binary(level - 1, ignore).unwrap();
        let ignore = ignore || Parser::order_of_operation(Plus) == level;

        let mut order = Parser::order_of_operation(self.current.token);
        while order == level {
            if ignore || self.lexer.peek(0).unwrap() == ' ' || self.lexer.peek(-2).unwrap() != ' ' {
                let op = self.current.token.clone();
                self.eat(op);
                let right = self.rec_binary(level - 1, ignore).unwrap();
                node = Expr::Binary(Box::new(Binary { left: node.clone(), op, right }))
            } else {
                return Ok(node);
            }
            order = Parser::order_of_operation(self.current.token);
        }

        return Ok(node);
    }

    fn expr(&mut self, ignore_white_space: bool) -> Result<Expr, Error> {
        let expr = self.rec_binary(OP_LEVELS, ignore_white_space).unwrap();

        if self.current.token == TokenType::Ternary {
            self.eat(TokenType::Ternary);
            let t = self.expr(ignore_white_space).unwrap();
            self.eat(Colon);
            let f = self.expr(ignore_white_space).unwrap();
            return Ok(Expr::Ternary(Box::new(Ternary { condition: expr, t_expr: t, f_expr: f })));
        }

        return Ok(expr);
    }

    fn assignments(&mut self) -> Result<Expr, Error> {
        let mut assignments = Vec::new();
        loop {
            let assignment = self.assignment().unwrap();
            assignments.push(assignment);

            if self.lexer.peek_token().unwrap().token == Comma {
                self.eat(Comma);
            } else {
                break;
            }
        }

        return Ok(Expr::Block(Box::new(Block { children: assignments })));
    }

    fn assignment(&mut self) -> Result<Expr, Error> {
        let left = self.variable().unwrap();

        self.eat(Assign);
        let right = self.expr(true).unwrap();

        return Ok(Expr::Assign(Box::new(Assignment { id: left, expr: right })));
    }

    fn variable(&mut self) -> Result<Expr, Error> {
        return if let TokenValue::String(id) = &self.current.value {
            let node = Expr::Var(Variable { id: id.clone() });
            self.eat(Id);
            Ok(node)
        } else {
            Err(Error::IdentifierNotFound(String::from("incorrect identifier")))
        };
    }

    fn factor(&mut self, ignore: bool) -> Result<Expr, Error> {
        let token = self.current.clone();
        return match token.token {
            TokenType::Plus => {
                self.eat(Plus);
                Ok(Expr::Unary(Box::new(Unary { op: token.token.clone(), expr: self.factor(ignore).unwrap() })))
            }
            TokenType::Minus => {
                self.eat(Minus);
                Ok(Expr::Unary(Box::new(Unary { op: token.token.clone(), expr: self.factor(ignore).unwrap() })))
            }
            TokenType::NotUnary => {
                self.eat(NotUnary);
                Ok(Expr::Unary(Box::new(Unary { op: token.token.clone(), expr: self.factor(ignore).unwrap() })))
            }
            TokenType::Num => {
                self.eat(Num);
                let result = &token.value;
                let mut data = Data::num(-1f64);
                match result {
                    TokenValue::Number(n) => {
                        data = Data::num(*n);
                    }
                    TokenValue::String(str) => {
                        data = Data::str(str.clone());
                    }
                }
                Ok(Expr::Val(Value { val: data }))
            }
            TokenType::LParen => {
                self.eat(LParen);
                let node = self.expr(ignore);
                self.eat(RParen);
                node
            }
            TokenType::LBracket => {
                Ok(Expr::Matrix(Box::new(self.matrix().unwrap())))
            }
            TokenType::ReservedValue => {
                Ok(self.reserved_value().unwrap())
            }
            TokenType::Id => {
                let next = self.lexer.peek_token().unwrap();
                if next.token == LParen {
                    let node = self.function_call().unwrap();
                    if self.current.token == Assign || self.current.token == Def {
                        return if let Expr::FnCall(node) = node {
                            self.function_definition(*node)
                        } else {
                            Err(Error::NotImplemented)
                        };
                    }
                    Ok(node)
                } else {
                    self.variable()
                }
            }
            _ => {
                Ok(Expr::Empty(Empty {}))
            }
        };
    }

    fn function_definition(&mut self, fn_call: FnCall) -> Result<Expr, Error> {
        self.eat(self.current.token);

        let mut args = Vec::new();
        for arg in &fn_call.args.seq {
            args.push(arg.clone());
        }

        if self.current.token == LBrace {
            // multi line statements
            self.eat(LBrace);
            let exprs = self.block().unwrap();
            self.eat(RBrace);
            return if let Expr::Block(block) = exprs {
                Ok(Expr::FnDecl(Box::new(FnDecl { id: fn_call.id.clone(), args: Sequence { seq: args }, func: *block })))
            } else {
                Err(Error::WrongType(String::from("Couldn't parse")))
            };
        } else {
            let expr = self.expr(true).unwrap();
            return Ok(Expr::FnDecl(Box::new(FnDecl { id: fn_call.id.clone(), args: Sequence { seq: args }, func: Block { children: vec![expr] } })));
        }

        return Err(Error::WrongType(String::from("Expected {")));
    }

    fn function_call(&mut self) -> Result<Expr, Error> {
        let name = self.current.value.clone();

        if let TokenValue::String(str) = name {
            self.eat(Id);
            self.eat(LParen);

            let mut args = Vec::new();
            while self.current.token != RParen {
                args.push(self.expr(true).unwrap());

                if self.current.token == Comma {
                    self.eat(Comma);
                } else {
                    break;
                }
            }
            self.eat(RParen);
            return Ok(Expr::FnCall(Box::new(FnCall { id: str.clone(), args: Sequence { seq: args } })));
        }

        return Err(Error::WrongType(String::from("Expected string identfiier")));
    }

    fn matrix(&mut self) -> Result<MatrixExpr, Error> {
        let mut arr = Vec::new();
        self.eat(LBracket);
        while self.current.token != RBracket {
            arr.push(self.matrix_row(TokenType::RBracket).unwrap());
            if self.current.token == Semicolon {
                self.eat(Semicolon);
            }
        }

        self.eat(RBracket);
        return Ok(MatrixExpr { matrix: arr });
    }

    fn matrix_row(&mut self, end: TokenType) -> Result<Vec<Expr>, Error> {
        let mut arr = Vec::new();

        loop {
            let expr = self.expr(false).unwrap();
            arr.push(expr);

            if self.current.token != Semicolon && self.current.token != end
            {
                if self.current.token == Comma {
                    self.eat(Comma);
                }
            } else {
                break;
            }
        }

        return Ok(arr);
    }

    fn reserved_value(&mut self) -> Result<Expr, Error> {
        let val = &self.current.value;

        if let TokenValue::String(val) = val {
            if val == "true" {
                self.eat(ReservedValue);
                return Ok(Expr::Val(Value { val: Data::bool(true) }));
            } else if val == "false" {
                self.eat(ReservedValue);
                return Ok(Expr::Val(Value { val: Data::bool(false) }));
            } else {
                return Err(Error::WrongType(String::from("Expected a boolean")));
            }
        }

        return Err(Error::WrongType(String::from("Expected a boolean")));
    }
    fn eat_new_lines(&mut self) {
        while self.current.token == EndL {
            self.eat(EndL);
        }
    }

    fn eat(&mut self, tok: TokenType) -> Option<Error> {
        // verify that token matches
        return if self.current.token == tok {
            self.current = self.lexer.next_token().unwrap();
            None
        } else {
            Some(Error::WrongType(String::from("Expected todo")))
        };
    }

    fn order_of_operation(token: TokenType) -> i8 {
        return match token {
            TokenType::Pow => { 0 }
            TokenType::Mod => { 1 }
            TokenType::Mul => { 2 }
            TokenType::Div => { 2 }
            TokenType::Plus => { 3 }
            TokenType::Minus => { 3 }
            TokenType::AndBit => { 4 }
            TokenType::OrBit => { 4 }
            TokenType::LessEq => { 5 }
            TokenType::MoreEq => { 5 }
            TokenType::RArrow => { 5 }
            TokenType::LArrow => { 5 }
            TokenType::Eq => { 5 }
            TokenType::NotEq => { 5 }
            TokenType::AndBool => { 6 }
            TokenType::OrBool => { 7 }
            _ => { i8::MIN }
        };
    }
}

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use crate::ast::{compute, Evaluates};
    use crate::symtable::SymTable;
    use crate::interpreter::interpret;


    fn ast(text: &str) {
        let mut parser = Parser::new(text);
        let expr = parser.parse().unwrap();
        println!("{:#?}", expr);
    }

    #[test]
    fn test_addition() {
        let result = interpret("1.2 + (1 - 10 / 2) * 6");
        println!("{:#?}", result);
    }

    #[test]
    fn test_inequality() {
        let result = interpret("1 + 3 > 4");
        println!("{:#?}", result);
    }

    #[test]
    fn test_unary() {
        let result = interpret("-(3+9)");
        println!("{:#?}", result);
    }

    #[test]
    fn test_ternary() {
        let result = interpret("x = 3\ny = x + 2\ny = y - 2\nb = x == y ? 2 : 3\nb + 4");
        println!("{:#?}", result);
    }

    #[test]
    fn test_matrix() {
        let result = interpret("[1 2 3 ] * [1; 2; 3] / 10");
        println!("{:#?}", result);
    }

    #[test]
    fn test_fn_decl() {
        let result = interpret("f(x, y, z) => x ^ y + z\nf(3, 6.5, 0)");
        println!("{:#?}", result);
    }

    #[test]
    fn test_recursion() {
        let result = interpret("fact(n) => {\nn == 0 ? 1 : n * fact(n - 1)\n}\nx = fact(5)\nx+90");
        println!("{:#?}", result);
    }

    #[test]
    fn test_err() {
        let result = interpret("x = 3\ny + 2");
        println!("{:#?}", result);
    }
}